pkg/controller/job/job_controller.go

```diff
@@ -1030,15 +1030,26 @@ func (jm *Controller) trackJobStatusAndRemoveFinalizers(ctx context.Context, job
 	podFailureCountByPolicyAction := map[string]int{}
 	for _, pod := range pods {
 		if !hasJobTrackingFinalizer(pod) || expectedRmFinalizers.Has(string(pod.UID)) {
+			// This pod was processed in a previous sync.
 			continue
 		}
-		podFinished := pod.Status.Phase == v1.PodSucceeded || pod.Status.Phase == v1.PodFailed
 		// Terminating pods are counted as failed. This guarantees that orphan Pods
 		// count as failures.
 		// Active pods are terminated when the job has completed, thus they count as
 		// failures as well.
-		podTerminating := pod.DeletionTimestamp != nil || finishedCond != nil
-		if podFinished || podTerminating || job.DeletionTimestamp != nil {
+		considerTerminated := pod.DeletionTimestamp != nil || finishedCond != nil
+
+		if feature.DefaultFeatureGate.Enabled(features.PodDisruptionConditions) && feature.DefaultFeatureGate.Enabled(features.JobPodFailurePolicy) && job.Spec.PodFailurePolicy != nil {
+			// TODO(#113855): Stop limiting this behavior to Jobs with podFailurePolicy.
+			// For now, we do so to avoid affecting all running Jobs without the
+			// avaibility to opt-out into the old behavior.
+			// We can also simplify the check to remove finalizers to:
+			// considerTerminated || job.DeletionTimestamp != nil
+			considerTerminated = podutil.IsPodTerminal(pod) ||
+				finishedCond != nil || // The Job is terminating. Any running Pod is considered failed.
```

Comment:
IIUC this change is only loosely related to the issue. Once we start to consider pods as terminated when `finishedCond!=nil` then they will be checked against the pod failure policy. This seems to make this comment obsolete: https://github.com/kubernetes/kubernetes/blob/8e48df135318026d5f8a972a96a2ff665889568a/pkg/controller/job/pod_failure_policy.go#L97.

Reply:
I see. still, inside the `if` we can skip the OR with `podutil.IsPodTerminal(pod)` to be more consistent with what is outside, altough then the var name would not be very descriptive. I fine to keep as is or remove the extra condition. It should not impact the behavior anyway.

Reply:
Eventually, I would like this logic inside the conditional to be the logic for all jobs.
Keeping it like this seems like a good way to soak the logic.